pkg/registry/service/rest.go

```diff
@@ -265,6 +276,97 @@ func (*REST) NewList() runtime.Object {
+	//
+	oldServiceHasHealthCheckNodePort := shouldCheckOrAssignHealthCheckNodePort(oldService)
+	oldHealthCheckNodePort := apiservice.GetServiceHealthCheckNodePort(oldService)
+
+	assignHealthCheckNodePort := shouldCheckOrAssignHealthCheckNodePort(service)
+	requestedHealthCheckNodePort := apiservice.GetServiceHealthCheckNodePort(service)
+
+	switch {
+	case !oldServiceHasHealthCheckNodePort && assignHealthCheckNodePort:
+		glog.Infof("Transition from Global LB service to OnlyLocal service")
+		if requestedHealthCheckNodePort > 0 {
+			// If the request has a health check nodePort in mind, attempt to reserve it
+			err := rs.serviceNodePorts.Allocate(int(requestedHealthCheckNodePort))
+			if err != nil {
+				errmsg := fmt.Sprintf("Failed to allocate requested HealthCheck nodePort %v:%v",
+					requestedHealthCheckNodePort, err)
+				el := field.ErrorList{field.Invalid(field.NewPath("metadata", "annotations"),
+					apiservice.AnnotationHealthCheckNodePort, errmsg)}
+				return false, errors.NewInvalid(api.Kind("Service"), service.Name, el)
+			}
+			glog.Infof("Reserved user requested nodePort: %d", requestedHealthCheckNodePort)
+		} else {
+			// If the request has no health check nodePort specified, allocate any
+			healthCheckNodePort, err := rs.serviceNodePorts.AllocateNext()
+			if err != nil {
+				// TODO: what error should be returned here?  It's not a
+				// field-level validation failure (the field is valid), and it's
+				// not really an internal error.
+				return false, errors.NewInternalError(fmt.Errorf("failed to allocate a nodePort: %v", err))
+			}
+			// Insert the newly allocated health check port as an annotation (plan of record for Alpha)
+			service.Annotations[apiservice.AnnotationHealthCheckNodePort] = fmt.Sprintf("%d", healthCheckNodePort)
+			glog.Infof("Reserved health check nodePort: %d", healthCheckNodePort)
+		}
+
+	case oldServiceHasHealthCheckNodePort && !assignHealthCheckNodePort:
+		glog.Infof("Transition from OnlyLocal LB service to Global service")
+		err := rs.serviceNodePorts.Release(int(oldHealthCheckNodePort))
+		if err != nil {
+			glog.Warningf("Error releasing service health check %s node port %d: %v", service.Name, oldHealthCheckNodePort, err)
+			return false, errors.NewInternalError(fmt.Errorf("failed to free health check nodePort: %v", err))
+		} else {
+			delete(service.Annotations, apiservice.AnnotationHealthCheckNodePort)
+			glog.Infof("Freed health check nodePort: %d", oldHealthCheckNodePort)
+		}
+
+	case !oldServiceHasHealthCheckNodePort && !assignHealthCheckNodePort:
+		if _, ok := service.Annotations[apiservice.AnnotationHealthCheckNodePort]; ok {
+			glog.Warningf("Attempt to insert health check node port annotation DENIED")
+			el := field.ErrorList{field.Invalid(field.NewPath("metadata", "annotations"),
+				apiservice.AnnotationHealthCheckNodePort, "Cannot insert healthcheck nodePort annotation")}
+			return false, errors.NewInvalid(api.Kind("Service"), service.Name, el)
+		}
```

Comment:
worth adding a comment that this is what happens in the normal Service update case when the feature is gated (so every service on aws, azure etc).